libtransmission/variant.cc

```diff
@@ -819,45 +820,36 @@ static void nodeDestruct(struct SaveNode* node)
  */
 void tr_variantWalk(tr_variant const* v_in, struct VariantWalkFuncs const* walkFuncs, void* user_data, bool sort_dicts)
 {
-    int stackSize = 0;
-    int stackAlloc = 64;
-    struct SaveNode* stack = tr_new(struct SaveNode, stackAlloc);
+    auto stack = std::vector<WalkNode>{};
+    stack.emplace_back(v_in, sort_dicts);
 
-    nodeConstruct(&stack[stackSize++], v_in, sort_dicts);
-
-    while (stackSize > 0)
+    while (!stack.empty())
     {
-        struct SaveNode* node = &stack[stackSize - 1];
+        auto& node = stack.back();
         tr_variant const* v;
 
-        if (!node->isVisited)
+        if (!node.is_visited)
         {
-            v = node->v;
-            node->isVisited = true;
+            v = &node.v;
+            node.is_visited = true;
         }
-        else if (tr_variantIsContainer(node->v) && node->childIndex < node->v->val.l.count)
+        else if (tr_variantIsContainer(&node.v) && ((v = node.nextChild())))
```

Comment:
Use proper boolean expressions.
```suggestion
        else if (tr_variantIsContainer(&node.v) && (v = node.nextChild()) != nullptr)
```
or
```suggestion
        else if (v = tr_variantIsContainer(&node.v) ? node.nextChild() : nullptr; v != nullptr)
```

Reply:
The new safeguard in ba418d1 to make `node.nextChild()` safe for non-containers makes this easier: 6f20296935ccdab9ca0ab1f88e8470d803dfd7f8